### DISCUSSION TITLE
Port spark sql tests to datafusion.

### DISCUSSION THREAD

**Discussion by liurenjie1024** [apache/datafusion#9693] [Ideas]
Spark sql has been widely used in large scale etl, and we have see 
[comet](https://github.com/apache/arrow-datafusion-comet) donated to apache, which acts as an execution layer of spark sql. 
To further improve datafusion's compatibility with spark sql, I propose to introduce port [spark sql tests](https://github.com/apache/spark/blob/b01b88ad68a264396001d99c63bf58fb56d2be68/sql/core/src/test/resources/sql-tests). 
This will help to bring datafusion to next level in etl processing.

**Comment by Omega359** [comment]
Spark functions do not yet map well to DataFusion functions so if those are included in the tests those will be an issue. I'm hoping that at some point in the (near?) future once the function migration is complete a new epic could be created to align function names, params and behaviour with the sql dialect being used (primarily I would expect PostgreSQL and Spark). That would generally be a pretty massive project though as anyone who has tried to mirror the behaviour of a different system can attest to - just talk to @andygrove.

**Reply by andygrove** [reply]
I like the idea of having datafusion crates with functions that are compatible with different query engines, with Postgres and Spark being the two that I am most interested in. Spark-compatible functions are currently being developed in the comet project.

Comet is taking the approach of running all of Spark's tests against Spark with Comet enabled, including the SQL tests, so porting the SQL tests may be redundant and require more maintenance to keep them up to date with new tests being added to Spark, as well as maintaining different tests for different Spark versions. Perhaps @sunchao or @viirya could add their thoughts.

**Reply by viirya** [reply]
For Comet, one major goal is fully compatibility with Spark, so we care about the necessary tests to make sure that. As I replied in the related ticket https://github.com/apache/arrow-datafusion-comet/issues/123#issuecomment-1965975507, porting existing Spark SQL tests to sqllogictest seems redundant because we have been running all Spark SQL tests with Comet enabled to make sure the compatibility. Not to mention that there are extra development and maintenance cost around these duplicate tests.

The idea of making DataFusion itself compatible with Spark/Postgres is another story, I think. It is not just porting tests from other engines but also much work are necessary to make DataFusion configurable to behave like/compatible with other engines.

**Reply by ozankabak** [reply]
I agree. Maybe the right place for Spark SQL tests is Comet and there is not much to gain from duplicating those here. It sounds sensible to me to have standard SQL conformity as the main goal at the upstream DataFusion level and have subprojects specialize on that, adding the tests necessary.

**Reply by liurenjie1024** [reply]
https://github.com/apache/arrow-datafusion-comet/issues/123#issuecomment-1965975507 is a different story, and at that time I didn't realize that spark already has similar data driven tests.

I may need to clarify my goal here. I think comet is a proof that datafusion is production ready. I'm thinking why not pushing datafusion/ballista to totally replace spark sql in large scale etl world. Porting spark sql tests is just first step of the goal.

**Reply by alamb** [reply]
> I may need to clarify my goal here. I think comet is a proof that datafusion is production ready. I'm thinking why not pushing datafusion/ballista to totally replace spark sql in large scale etl world. Porting spark sql tests is just first step of the goal.

This is a noble goal, but it will require far more than just good code / features (docs, evangelism, etc) -- and is likely orthogonal to the goals of comet (aka it can be pursued in paralle)

**Reply by andygrove** [reply]
Another potential option here is for Comet to make spark-compatible operators and expressions available in a crate that Ballista and other projects could leverage, if they need Spark SQL support.

**Comment by alamb** [comment]
I think it is a great idea to use existing test suites to improve compatibility / correctness of DataFusion. Thank you @liurenjie1024

**Comment by ozankabak** [comment]
Let's just remain careful to follow standard SQL as default and not Spark SQL in terms of function names, syntaxes etc. We should obviously accept the latter if the user configures so, but there could be some work to do there to attain this level of flexibility.

**Reply by alamb** [reply]
I am not sure if there is any such thing as truly standard SQL

However, I think now DataFusion  strives to support the Postgres dialect /semantics. Keeping these semantics are also critical for us at InfluxData

So perhaps we can strive for "postgres compatible function package" and "spark compatible function packages" 🤔

**Reply by ozankabak** [reply]
Well there is ISO SQL, which is what I look at when in doubt :) PostgreSQL is quite compliant to that, it differs every now and then but it is a very good proxy.

**Reply by alamb** [reply]
There is a standard, but I think it is basically a running joke in database implementation circles as no database actually implements the whole thing faithfully

**Reply by liurenjie1024** [reply]
Yeah, it's hard to totally follow ansi sql, especially for distributed engine. For example, spark sql's create table statement has some extension for partition, clustering, and I'm not sure if it exists in ansi sql.

**Reply by alamb** [reply]
to be clear my point is not that we shouldn't try to follow postgres faithfully, but more like since there is no universally agreed upon SQL spec, we need to be prepared for people to need different dialects and datafusion should let them implement those dialects
